pkg/blockbuilder/blockbuilder.go

```diff
@@ -239,3 +240,308 @@ func (b *BlockBuilder) getLagForPartition(ctx context.Context, partition int32)
+		// Instead of looking for the commit metadata for each iteration, we use the data returned by consumePartition in the next iteration.
+		err := b.consumePartitionCycle(ctx, state, ce)
+		if err != nil {
+			return fmt.Errorf("consume partition %d: %w", state.Partition, err)
+		}
+
+		// If adding the ConsumeInterval takes it beyond the cycleEnd, we set it to the cycleEnd to not
+		// exit the loop without consuming until cycleEnd.
+		if ce.Compare(cycleEnd) != 0 && ce.Add(b.cfg.ConsumeInterval).After(cycleEnd) {
+			ce = cycleEnd
+		}
+	}
+
+	return nil
+}
+
+// consumePartition consumes records from the given partition until the cycleEnd timestamp.
+// It modifies state, tracking the consumption progress. If the partition is lagging behind, the caller needs to take care
+// of calling consumePartition in sections.
+func (b *BlockBuilder) consumePartitionCycle(ctx context.Context, state *partitionState, cycleEnd time.Time) (retErr error) {
+	blockEndAt := cycleEnd.Truncate(b.cfg.ConsumeInterval)
+	blockEnd := blockEndAt.UnixMilli()
+
+	var (
+		numBlocks     int
+		compactionDur time.Duration
+	)
+	defer func(t time.Time, startState partitionState) {
+		dur := time.Since(t)
+
+		// Don't propagate any errors derived from the canceled context.
+		if errors.Is(retErr, context.Canceled) {
+			retErr = nil
+		}
+
+		if retErr != nil {
+			level.Error(b.logger).Log("msg", "partition consumption failed", "partition", state.Partition, "dur", dur, "start_lag", startState.Lag, "end_lag", state.Lag, "err", retErr)
+			return
+		}
+
+		b.metrics.processPartitionDuration.WithLabelValues(fmt.Sprintf("%d", state.Partition)).Observe(dur.Seconds())
+		level.Info(b.logger).Log("msg", "done consuming", "partition", state.Partition, "dur", dur,
+			"start_lag", startState.Lag, "end_lag", state.Lag, "cycle_end", cycleEnd,
+			"last_block_end", time.UnixMilli(startState.LastBlockEnd), "curr_block_end", time.UnixMilli(state.LastBlockEnd),
+			"last_seen_till", startState.SeenTillOffset, "curr_seen_till", state.SeenTillOffset,
+			"num_blocks", numBlocks, "compact_and_upload_dur", compactionDur)
+	}(time.Now(), *state)
+
+	builder := NewTSDBBuilder(b.logger, b.cfg.DataDir, b.cfg.BlocksStorage, b.limits)
+	defer builder.Close()
+
+	level.Info(b.logger).Log("msg", "start consuming", "partition", state.Partition, "offset", state.Commit.At, "lag", state.Lag, "cycle_end", cycleEnd)
+
+	// TopicPartition to resume consuming on this iteration.
+	tp := map[string][]int32{b.cfg.Kafka.Topic: {state.Partition}}
+	b.kafkaClient.ResumeFetchPartitions(tp)
+	defer b.kafkaClient.PauseFetchPartitions(tp)
+
+	// We always rewind the partition's offset to the commit offset. This is so the cycle started exactly at the commit point,
+	// and not at what was (potentially over-) consumed previously.
+	b.seekPartition(state.Partition, state.Commit.At)
+
+	var (
+		firstRec  *kgo.Record
+		lastRec   *kgo.Record
+		commitRec *kgo.Record
```

Comment:
In order to commit the state, we need two pieces:

1. offset of the record (`kgo.Record.Offset`)
2. timestamp of the record (`kgo.Record.Timestamp.UnixMilli()`)

I don't think there is a single struct in `kgo` that has both of those, other than the whole `Record`.

Reply:
ok, then what about the naming?

Reply:
>  I feel like some name such as `firstPartiallyProcessedRecord` would be more obvious.

I've spent too much time debating about that with myself already. The `commitRec` points at the record whose offset we want to commit.

- if we fully processed all records in this cycle, this will be the last consumed record
- otherwise, it will be the **last** fully processed record before the first partially processed one 🤯
- or, if the very first record was partially processed, it points to the previous commit (updating the commit's metadata).

I'm fine renaming it, but this name doesn't make sense to me.

Reply:
let's keep it as-is